File: Strings/SortCharactersByFrequency.cpp
// https://leetcode.com/problems/sort-characters-by-frequency/

#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
#define FOR(a, b, c) for (int(a) = (b); (a) < (c); ++(a))
#define FORD(a, b, c) for (int(a) = (b); (a) >= (c); --(a))
#define nl cout << "\n";
#define vi vector<int>
#define vi vector<int>
#define vs vector<string>
#define pb push_back
#define vvi vector<vector<int>>
#define umap unordered_map

string frequencySort(string s)
{
    umap<char, int> map;
    FOR(i, 0, s.length())
    {
        map[s[i]]++;
    }
    string res = "";
    vector<pair<int, char>> A(map.size());
    for (auto it : map)
    {
        A.push_back({it.second, it.first});
    }
    sort(A.rbegin(), A.rend());

    for (auto x : A)
    {
        FOR(i, 0, x.first)
        res += x.second;
    }
    return res;
}

int main()
{

    string s = "eeeeaaarrd";
    cout << frequencySort(s);
}

File: LeetcodeDaily/SortCharactersByFrequency.java
// https://leetcode.com/problems/sort-characters-by-frequency/description/

import java.util.*;

public class SortCharactersByFrequency {
    
    class Pair{
        int first ;  char second;
        Pair (int a , char b){
            first = a;
            second = b;
        }
    }
    class SortByFreq implements Comparator<Pair> {  
        // Used for sorting in ascending order of ID  
        public int compare(Pair i1, Pair i2)  
        {  
            if (i1.first == i2.first)
            return 0;
            else if (i1.first < i2.first)
            return 1;
            else
            return -1;
        }  
       
    }  
    public String frequencySort(String s) {
        StringBuilder res = new StringBuilder();
        HashMap<Character,Integer> map = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            if(map.containsKey(s.charAt(i))){
                map.put(s.charAt(i), map.get(s.charAt(i))+1);
            }
            else{
                map.put(s.charAt(i) , 1 );
            }
        }

        List<Pair> lst = new ArrayList<>();
        map.forEach((k,v) -> {
            lst.add(new Pair(v, k));
        });
        
        Collections.sort(lst, new SortByFreq());  
        for (Pair pair : lst) {

            for (int i = 0; i < pair.first; i++) {
                res.append(pair.second);
            }

        }
        return res.toString();
    
    }
}